src/core_plugins/input_control_vis/public/components/editor/field_select.js

```diff
@@ -1,60 +1,108 @@
 import _ from 'lodash';
 import PropTypes from 'prop-types';
 import React, { Component } from 'react';
-import Select from 'react-select';
 
 import {
   EuiFormRow,
+  EuiComboBox,
 } from '@elastic/eui';
 
 export class FieldSelect extends Component {
   constructor(props) {
     super(props);
 
-    // not storing activeIndexPatternId in react state
-    // 1) does not effect rendering
-    // 2) requires synchronous modification to avoid race condition
-    this.activeIndexPatternId = props.indexPatternId;
-
     this.state = {
-      fields: []
+      isLoading: false,
+      fields: [],
+      indexPatternId: props.indexPatternId,
     };
     this.filterField = _.get(props, 'filterField', () => { return true; });
-    this.loadFields(props.indexPatternId);
+  }
+
+  componentWillMount() {
+    this._isMounted = true;
+  }
+
+  componentWillUnmount() {
+    this._isMounted = false;
+  }
+
+  componentDidMount() {
+    this.loadFields(this.state.indexPatternId);
   }
 
   componentWillReceiveProps(nextProps) {
     if (this.props.indexPatternId !== nextProps.indexPatternId) {
-      this.activeIndexPatternId = nextProps.indexPatternId;
-      this.setState({ fields: [] });
       this.loadFields(nextProps.indexPatternId);
     }
   }
 
-  async loadFields(indexPatternId) {
+  loadFields = (indexPatternId) => {
+    this.setState({
+      isLoading: true,
+      fields: [],
+      indexPatternId
+    }, this.debouncedLoad.bind(null, indexPatternId));
+  }
+
+  debouncedLoad = _.debounce(async (indexPatternId) => {
     if (!indexPatternId || indexPatternId.length === 0) {
       return;
     }
 
     const indexPattern = await this.props.getIndexPattern(indexPatternId);
 
+    if (!this._isMounted) {
+      return;
+    }
+
     // props.indexPatternId may be updated before getIndexPattern returns
     // ignore response when fetched index pattern does not match active index pattern
-    if (indexPattern.id !== this.activeIndexPatternId) {
+    if (indexPattern.id !== this.state.indexPatternId) {
       return;
     }
 
-    const fields = indexPattern.fields
+    const fieldsByTypeMap = new Map();
```

Comment:
I would prefer to leave this the way it is. It is easy enough to follow and avoids making additional calls to lodash that require (at least for me) looking up the API call documentation instead of just using built in javascript syntax.